src/main/java/appeng/util/Platform.java

```diff
@@ -1182,29 +1187,39 @@ public static long nanoTime()
 			retrieved = possible.getStackSize();
 		}
 
-		final double availablePower = energy.extractAEPower( retrieved, Actionable.SIMULATE, PowerMultiplier.CONFIG );
-
-		final long itemToExtract = Math.min( (long) ( availablePower + 0.9 ), retrieved );
+		final double energyFactor = Math.max( 1.0, cell.getChannel().transferFactor() );
+		final double availablePower = energy.extractAEPower( retrieved / energyFactor, Actionable.SIMULATE, PowerMultiplier.CONFIG );
+		final long itemToExtract = Math.min( (long) ( ( availablePower * energyFactor ) + 0.9 ), retrieved );
 
 		if( itemToExtract > 0 )
 		{
-			energy.extractAEPower( retrieved, Actionable.MODULATE, PowerMultiplier.CONFIG );
-
-			possible.setStackSize( itemToExtract );
-			final T ret = cell.extractItems( possible, Actionable.MODULATE, src );
+			if( mode == Actionable.MODULATE )
+			{
+				energy.extractAEPower( retrieved / energyFactor, Actionable.MODULATE, PowerMultiplier.CONFIG );
+				possible.setStackSize( itemToExtract );
+				final T ret = cell.extractItems( possible, Actionable.MODULATE, src );
 
-			if( ret != null )
+				if( ret != null )
+				{
+					src.player().ifPresent( player -> Stats.ItemsExtracted.addToPlayer( player, (int) ret.getStackSize() ) );
+				}
+				return ret;
+			}
+			else
 			{
-				src.player().ifPresent( player -> Stats.ItemsExtracted.addToPlayer( player, (int) ret.getStackSize() ) );
+				return possible.setStackSize( itemToExtract );
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "possible" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2259)